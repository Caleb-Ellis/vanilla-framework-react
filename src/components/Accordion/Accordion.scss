@import 'vanilla-framework/scss/patterns_accordion';
@include vf-p-accordion;