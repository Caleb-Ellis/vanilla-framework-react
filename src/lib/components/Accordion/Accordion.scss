@import 'vanilla-framework/scss/base';
@import 'vanilla-framework/scss/patterns_accordion';
@include vf-b-typography;
@include vf-b-button;
@include vf-p-accordion;

.p-accordion__panel > :first-child {
  margin-top: 0;
}
