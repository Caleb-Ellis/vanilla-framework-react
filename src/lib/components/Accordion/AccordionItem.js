import React from 'react';
import PropTypes from 'prop-types';

const AccordionItem = (props) => {
  const {
    isOpen, onClick, index, title, children,
  } = props;

  return (
    <li className="p-accordion__group">
      <button
        className="p-accordion__tab"
        id={`accordion-button-${index}`}
        role="tab"
        aria-controls={`accordion-content-${index}`}
        aria-expanded={isOpen}
        onClick={() => onClick(index)}
        onKeyDown={key => key === 'Enter' && onClick(index)}
      >
        { title }
      </button>
      <section
        className="p-accordion__panel"
        id={`accordion-content-${index}`}
        role="tabpanel"
        aria-hidden={!isOpen}
        aria-labelledby={`accordion-button-${index}`}
      >
        { children }
      </section>
    </li>
  );
};

AccordionItem.defaultProps = {
  index: 0,
  isOpen: false,
  onClick: () => 1,
};

AccordionItem.propTypes = {
  index: PropTypes.number,
  title: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
  isOpen: PropTypes.bool,
  onClick: PropTypes.func,
};

AccordionItem.displayName = 'AccordionItem';

export default AccordionItem;
