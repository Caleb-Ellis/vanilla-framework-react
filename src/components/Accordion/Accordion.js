import React from 'react';
import AccordionItem from './AccordionItem';
import './Accordion.css';

class Accordion extends React.Component {
  render() {
    return (
      <aside className="p-accordion" role="tablist">
        <ul className="p-accordion__list">
          <AccordionItem title="Title of Item 1">
            <p>Lorem Ipsum has been the industrys standard dummy text ever since
            the 1500s, when an unknown printer took a galley of type and
            scrambled it to make a type specimen book.
            </p>
          </AccordionItem>
          <AccordionItem title="Title of Item 2">
            <p>Lorem Ipsum has been the industrys standard dummy text ever since
            the 1500s, when an unknown printer took a galley of type and
            scrambled it to make a type specimen book.
            </p>
          </AccordionItem>
          <AccordionItem title="Title of Item 3">
            <p>Lorem Ipsum has been the industrys standard dummy text ever since
            the 1500s, when an unknown printer took a galley of type and
            scrambled it to make a type specimen book.
            </p>
          </AccordionItem>
        </ul>
      </aside>
    );
  }
}

export default Accordion;
