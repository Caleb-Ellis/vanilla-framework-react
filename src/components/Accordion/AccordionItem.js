import React from 'react';
import PropTypes from 'prop-types';
import './Accordion.css';

class AccordionItem extends React.Component {
  constructor(props) {
    super(props);
    this.accordionItemToggle = this.accordionItemToggle.bind(this);
    this.state = {
      hidden: true,
    };
  }
  accordionItemToggle() {
    const currentState = this.state.hidden;
    this.setState({ hidden: !currentState });
  }
  render() {
    return (
      <li
        className="p-accordion__group"
      >
        <button
          className="p-accordion__tab"
          id="status-tab"
          role="tab"
          aria-controls="#status"
          aria-expanded={!this.state.hidden}
          onClick={() => this.accordionItemToggle(this)}
          onKeyDown={() => this.accordionItemToggle(this)}
        >
          { this.props.title }
        </button>
        <section
          className="p-accordion__panel"
          id="owner"
          role="tabpanel"
          aria-hidden={this.state.hidden}
          aria-labelledby="owner-tab"
        >
          { this.props.children }
        </section>
      </li>
    );
  }
}

AccordionItem.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
};

export default AccordionItem;
