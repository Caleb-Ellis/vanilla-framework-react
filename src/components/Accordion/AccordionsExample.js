import React from 'react';
import Accordion from './Accordion';

class AccordionsExample extends React.Component {
  render() {
    return (
      <div>
        <Accordion />
      </div>
    );
  }
}

export default AccordionsExample;
