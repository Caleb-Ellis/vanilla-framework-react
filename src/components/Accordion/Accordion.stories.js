import React from 'react';
import { storiesOf } from '@storybook/react';

import Accordion from './Accordion';
import AccordionItem from './AccordionItem';

storiesOf('Accordion', module)
  .add('Default', () => (
    <Accordion>
      <AccordionItem title="Title of Item 1">
        <p>Lorem Ipsum has been the industrys standard dummy text ever since
        the 1500s, when an unknown printer took a galley of type and
        scrambled it to make a type specimen book.
        </p>
      </AccordionItem>
      <AccordionItem title="Title of Item 2">
        <p>Lorem Ipsum has been the industrys standard dummy text ever since
        the 1500s, when an unknown printer took a galley of type and
        scrambled it to make a type specimen book.
        </p>
      </AccordionItem>
      <AccordionItem title="Title of Item 3">
        <p>Lorem Ipsum has been the industrys standard dummy text ever since
        the 1500s, when an unknown printer took a galley of type and
        scrambled it to make a type specimen book.
        </p>
      </AccordionItem>
    </Accordion>
  ));
