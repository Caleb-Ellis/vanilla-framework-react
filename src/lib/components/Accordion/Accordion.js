import React from 'react';
import PropTypes from 'prop-types';

import AccordionItem from './AccordionItem';

class Accordion extends React.Component {
  constructor(props) {
    super(props);
    this.toggleSection = this.toggleSection.bind(this);

    this.state = { open: props.selected };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ open: nextProps.selected });
  }

  toggleSection(index) {
    const { allowMultiple } = this.props;
    let newOpenArray = [...this.state.open];

    if (newOpenArray.includes(index)) {
      newOpenArray = newOpenArray.filter(item => item !== index);
    } else {
      newOpenArray = allowMultiple ? [...newOpenArray, index] : [index];
    }

    this.setState({ open: newOpenArray });
  }

  render() {
    const { open } = this.state;
    const accordionItems = React.Children.map(this.props.children,
      (child, index) => React.cloneElement(child, {
        index,
        isOpen: open.includes(index),
        onClick: this.toggleSection,
      }),
    );

    return (
      <aside className="p-accordion" role="tablist">
        <ul className="p-accordion__list">
          { accordionItems }
        </ul>
      </aside>
    );
  }
}

Accordion.defaultProps = {
  children: null,
  selected: [],
  allowMultiple: false,
};

Accordion.propTypes = {
  children: (props, propName, componentName) => {
    const prop = props[propName];
    let error = null;

    React.Children.forEach(prop, (child) => {
      if (child.type !== AccordionItem) {
        error = new Error(`${componentName} children should be of type "AccordionItem".`);
      }
    });

    return error;
  },
  selected: PropTypes.arrayOf(PropTypes.number),
  allowMultiple: PropTypes.bool,
};

Accordion.displayName = 'Accordion';

export default Accordion;
