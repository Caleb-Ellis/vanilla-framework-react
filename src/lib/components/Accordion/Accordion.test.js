import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Accordion from './Accordion';
import AccordionItem from './AccordionItem';

describe('<Accordion>', () => {
  it('renders Accordion with AccordionItems correctly', () => {
    const accordion = ReactTestRenderer.create(
      <Accordion>
        <AccordionItem title="Title of Item 1">
          <p>Lorem Ipsum has been the industrys standard dummy text ever since
          the 1500s, when an unknown printer took a galley of type and
          scrambled it to make a type specimen book.
          </p>
        </AccordionItem>
        <AccordionItem title="Title of Item 2">
          <p>Lorem Ipsum has been the industrys standard dummy text ever since
          the 1500s, when an unknown printer took a galley of type and
          scrambled it to make a type specimen book.
          </p>
        </AccordionItem>
        <AccordionItem title="Title of Item 3">
          <p>Lorem Ipsum has been the industrys standard dummy text ever since
          the 1500s, when an unknown printer took a galley of type and
          scrambled it to make a type specimen book.
          </p>
        </AccordionItem>
      </Accordion>,
    );
    const json = accordion.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('opens an AccordionItem when clicked', () => {
    const accordion = mount(
      <Accordion>
        <AccordionItem title="Accordion Title">
          <p>Accordion Content</p>
        </AccordionItem>
      </Accordion>);

    expect(accordion.find('[aria-expanded=true]')).toHaveLength(0);
    expect(accordion.find('[aria-hidden=true]')).toHaveLength(1);

    accordion.find('button').hostNodes().simulate('click');
    expect(accordion.find('[aria-expanded=true]')).toHaveLength(1);
    expect(accordion.find('[aria-hidden=true]')).toHaveLength(0);
  });

  it('only has maximum one AccordionItem open at a time if allowMultiple=false', () => {
    const accordion = mount(
      <Accordion>
        <AccordionItem title="Accordion Title 1">
          <p>Accordion Content 1</p>
        </AccordionItem>
        <AccordionItem title="Accordion Title 2">
          <p>Accordion Content 2</p>
        </AccordionItem>
      </Accordion>);

    expect(accordion.find('[aria-expanded=true]')).toHaveLength(0);

    accordion.find('#accordion-button-0').hostNodes().simulate('click');
    expect(accordion.find('[aria-expanded=true]')).toHaveLength(1);

    accordion.find('#accordion-button-1').hostNodes().simulate('click');
    expect(accordion.find('[aria-expanded=true]')).toHaveLength(1);

    accordion.find('#accordion-button-1').hostNodes().simulate('click');
    expect(accordion.find('[aria-expanded=true]')).toHaveLength(0);
  });

  it('allows multiple AccordionItems open at a time if allowMultiple=true', () => {
    const accordion = mount(
      <Accordion allowMultiple>
        <AccordionItem title="Accordion Title 1">
          <p>Accordion Content 1</p>
        </AccordionItem>
        <AccordionItem title="Accordion Title 2">
          <p>Accordion Content 2</p>
        </AccordionItem>
      </Accordion>);

    expect(accordion.find('[aria-expanded=true]')).toHaveLength(0);

    accordion.find('#accordion-button-0').hostNodes().simulate('click');
    expect(accordion.find('[aria-expanded=true]')).toHaveLength(1);

    accordion.find('#accordion-button-1').hostNodes().simulate('click');
    expect(accordion.find('[aria-expanded=true]')).toHaveLength(2);

    accordion.find('#accordion-button-1').hostNodes().simulate('click');
    expect(accordion.find('[aria-expanded=true]')).toHaveLength(1);
  });

  it('allows pre-selection of AccordionItems if selected prop provided', () => {
    const accordion = mount(
      <Accordion selected={[0]}>
        <AccordionItem title="Accordion Title 1">
          <p>Accordion Content 1</p>
        </AccordionItem>
        <AccordionItem title="Accordion Title 2">
          <p>Accordion Content 2</p>
        </AccordionItem>
      </Accordion>);

    expect(accordion.find('[aria-expanded=true]')).toHaveLength(1);
  });
});

