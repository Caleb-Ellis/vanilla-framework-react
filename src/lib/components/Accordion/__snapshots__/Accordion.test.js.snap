// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion> renders Accordion with AccordionItems correctly 1`] = `
<aside
  className="p-accordion"
  role="tablist"
>
  <ul
    className="p-accordion__list"
  >
    <li
      className="p-accordion__group"
    >
      <button
        aria-controls="accordion-content-0"
        aria-expanded={false}
        className="p-accordion__tab"
        id="accordion-button-0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
      >
        Title of Item 1
      </button>
      <section
        aria-hidden={true}
        aria-labelledby="accordion-button-0"
        className="p-accordion__panel"
        id="accordion-content-0"
        role="tabpanel"
      >
        <p>
          Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
        </p>
      </section>
    </li>
    <li
      className="p-accordion__group"
    >
      <button
        aria-controls="accordion-content-1"
        aria-expanded={false}
        className="p-accordion__tab"
        id="accordion-button-1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
      >
        Title of Item 2
      </button>
      <section
        aria-hidden={true}
        aria-labelledby="accordion-button-1"
        className="p-accordion__panel"
        id="accordion-content-1"
        role="tabpanel"
      >
        <p>
          Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
        </p>
      </section>
    </li>
    <li
      className="p-accordion__group"
    >
      <button
        aria-controls="accordion-content-2"
        aria-expanded={false}
        className="p-accordion__tab"
        id="accordion-button-2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
      >
        Title of Item 3
      </button>
      <section
        aria-hidden={true}
        aria-labelledby="accordion-button-2"
        className="p-accordion__panel"
        id="accordion-content-2"
        role="tabpanel"
      >
        <p>
          Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
        </p>
      </section>
    </li>
  </ul>
</aside>
`;
